### PR TITLE
Use `esbuild` to bundle some CommonJS main files

### PR DESCRIPTION
**Pull request by bpasero**

### CODE REVIEW

File: build/gulpfile.editor.js [hunk 1]
```diff
@@ -40,7 +40,6 @@ const editorEntryPoints = [
 		exclude: ['vs/nls'],
 		prepend: [
 			{ path: 'vs/loader.js' },
-			{ path: 'vs/nls.js', amdModuleId: 'vs/nls' },
```
Comment (bpasero) [inline build/gulpfile.editor.js:43]: @alexdima not so sure about this one here

Comment (alexdima) [inline build/gulpfile.editor.js:43]: Indeed, I think this needs to stay. Even after adding this line back, when I run locally `yarn gulp editor-distro`, it now fails with the following:
```
[09:27:49] The following tasks did not complete: editor-distro
[09:27:49] Did you forget to signal async completion?
error Command failed with exit code 1.
```
This almost looks like we forget to return a Promise or a stream somewhere

File: build/gulpfile.vscode.web.js [hunk 2]
```diff
@@ -153,24 +153,26 @@ exports.createVSCodeWebFileContentMapper = createVSCodeWebFileContentMapper;
...
-		otherSources: [],
-		resources: vscodeWebResources,
-		loaderConfig: common.loaderConfig(),
-		externalLoaderInfo: util.createExternalLoaderConfig(product.webEndpointUrl, commit, quality),
-		out: 'out-vscode-web',
-		inlineAmdImages: true,
-		bundleInfo: undefined,
-		fileContentMapper: createVSCodeWebFileContentMapper('.build/web/extensions', product)
-	})
+	optimize.optimizeTask(
```
Comment (alexdima) [inline build/gulpfile.vscode.web.js:156]: This call seems all broken. I would have expected again the format `{ out: '...', amd: { ... } }`

**Review by alexdima — CHANGES_REQUESTED** [review]
Noticed a couple of problems. I'm investigating locally why `editor-distro` would now be broken.

File: build/lib/optimize.ts [hunk 3]
```diff
@@ -205,67 +201,123 @@ const DEFAULT_FILE_HEADER = [
...
+	out: string;
+	amd: IOptimizeAMDTaskOpts;
+	commonJS?: IOptimizeCommonJSTaskOpts;
+}
+
+export function optimizeTask(opts: IOptimizeTaskOpts): () => NodeJS.ReadWriteStream {
+	return function () {
+		return es.merge(
+			optimizeAMDTask(opts.amd),
+			opts.commonJS ? optimizeCommonJSTask(opts.commonJS) : es.through()
```
Comment (alexdima) [inline build/lib/optimize.ts:319]: `es.merge( something, es.through() )` is the root cause for `yarn gulp editor-distro` failing. We could do `es.merge` over a spread array and conditionally push to the array before if necessary.

File: build/gulpfile.vscode.web.js (hunk 2 above)
Comment (bpasero) [reply build/gulpfile.vscode.web.js:156]: https://github.com/microsoft/vscode/pull/160957/commits/bcbac86f7a521d4cd62ca86c2844e60f1acfe0f9

File: build/lib/optimize.ts (hunk 3 above)
Comment (bpasero) [reply build/lib/optimize.ts]: Using spread array https://github.com/microsoft/vscode/pull/160957/commits/f5e095bc4d74b6969d55a55b35487ff048f84be5

File: build/gulpfile.editor.js (hunk 1 above)
Comment (bpasero) [reply build/gulpfile.editor.js:43]: Maybe fixed with https://github.com/microsoft/vscode/pull/160957/commits/f5e095bc4d74b6969d55a55b35487ff048f84be5?

**Review by alexdima — APPROVED** [review]
Looks good!

**Pull request merged**